File: emulators/emulator-base-native/c/mle_nativearchive.c
/* -*- Mode: C++; indent-tabs-mode: t; tab-width: 4 -*-
// ---------------------------------------------------------------------------
// Multi-Phasic Applications: SquirrelJME
//	 Copyright (C) Stephanie Gawroriski <xer@multiphasicapps.net>
// ---------------------------------------------------------------------------
// SquirrelJME is under the Mozilla Public License Version 2.0.
// See license.mkd for licensing and copyright information.
// ------------------------------------------------------------------------ */

#include "squirreljme.h"

#define NAME_ARCHIVE "cc/squirreljme/jvm/mle/brackets/NativeArchiveBracket"
#define NAME_ENTRY "cc/squirreljme/jvm/mle/brackets/NativeArchiveEntryBracket"

#define FORWARD_CLASS "cc/squirreljme/jvm/mle/NativeArchiveShelf"
#define FORWARD_CLASS_NAME NativeArchive
#define FORWARD_NATIVE_CLASS "cc/squirreljme/emulator/EmulatedNativeArchiveShelf"

#define FORWARD_DESC_archiveClose DESC_METHOD(DESC_VOID,  \
	DESC_CLASS(NAME_ARCHIVE))
#define FORWARD_DESC_archiveEntry DESC_METHOD(DESC_CLASS(NAME_ENTRY),  \
	DESC_CLASS(NAME_ARCHIVE) DESC_STRING)
#define FORWARD_DESC_archiveOpenZip DESC_METHOD(DESC_CLASS(NAME_ARCHIVE),  \
	DESC_ARRAY(DESC_BYTE) DESC_INT DESC_INT)
#define FORWARD_DESC_entryIsDirectory DESC_METHOD(DESC_BOOLEAN,  \
	DESC_CLASS(NAME_ENTRY))
#define FORWARD_DESC_entryOpen DESC_METHOD(DESC_INPUT_STREAM,  \
	DESC_CLASS(NAME_ENTRY))
#define FORWARD_DESC_entryUncompressedSize DESC_METHOD(DESC_LONG,  \
	DESC_CLASS(NAME_ENTRY))

FORWARD_IMPL_VOID(NativeArchive, archiveClose, \
	FORWARD_IMPL_args(jobject archive), \
	FORWARD_IMPL_pass(archive))
FORWARD_IMPL(NativeArchive, archiveEntry, jobject, Object, \
	FORWARD_IMPL_args(jobject archive, jobject name), \
	FORWARD_IMPL_pass(archive, name))
FORWARD_IMPL(NativeArchive, archiveOpenZip, jobject, Object, \
	FORWARD_IMPL_args(jobject buf, jint off, jint len), \
	FORWARD_IMPL_pass(buf, off, len))
FORWARD_IMPL(NativeArchive, entryIsDirectory, jboolean, Boolean, \
	FORWARD_IMPL_args(jobject entry), \
	FORWARD_IMPL_pass(entry))
FORWARD_IMPL(NativeArchive, entryOpen, jobject, Object, \
	FORWARD_IMPL_args(jobject entry), \
	FORWARD_IMPL_pass(entry))
FORWARD_IMPL(NativeArchive, entryUncompressedSize, jlong, Long, \
	FORWARD_IMPL_args(jobject entry), \
	FORWARD_IMPL_pass(entry))

static const JNINativeMethod mleNativeArchiveMethods[] =
{
	FORWARD_list(NativeArchive, archiveClose),
	FORWARD_list(NativeArchive, archiveEntry),
	FORWARD_list(NativeArchive, archiveOpenZip),
	FORWARD_list(NativeArchive, entryIsDirectory),
	FORWARD_list(NativeArchive, entryOpen),
	FORWARD_list(NativeArchive, entryUncompressedSize)
};

FORWARD_init(mleNativeArchiveInit, mleNativeArchiveMethods)

File: emulators/springcoat-vm/src/main/java/cc/squirreljme/vm/springcoat/MLENativeArchive.java
// -*- Mode: Java; indent-tabs-mode: t; tab-width: 4 -*-
// ---------------------------------------------------------------------------
// Multi-Phasic Applications: SquirrelJME
//     Copyright (C) Stephanie Gawroriski <xer@multiphasicapps.net>
// ---------------------------------------------------------------------------
// SquirrelJME is under the Mozilla Public License Version 2.0.
// See license.mkd for licensing and copyright information.
// ---------------------------------------------------------------------------

package cc.squirreljme.vm.springcoat;

import cc.squirreljme.jvm.mle.NativeArchiveShelf;
import cc.squirreljme.jvm.mle.brackets.NativeArchiveBracket;
import cc.squirreljme.jvm.mle.brackets.NativeArchiveEntryBracket;
import cc.squirreljme.jvm.mle.exceptions.MLECallError;
import cc.squirreljme.vm.springcoat.brackets.NativeArchiveEntryObject;
import cc.squirreljme.vm.springcoat.brackets.NativeArchiveObject;
import cc.squirreljme.vm.springcoat.exceptions.SpringMLECallError;
import java.io.IOException;

/**
 * SpringCoat layer for {@link NativeArchiveShelf}.
 *
 * @since 2024/03/05
 */
public enum MLENativeArchive
	implements MLEFunction
{
	/** {@link NativeArchiveShelf#archiveClose(NativeArchiveBracket)}. */
	ARCHIVE_CLOSE("archiveClose:" +
		"(Lcc/squirreljme/jvm/mle/brackets/NativeArchiveBracket;)V")
	{
		/**
		 * {@inheritDoc}
		 * @since 2024/03/05
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			Object archive = __args[0];
			
			try
			{
				NativeArchiveShelf.archiveClose(
					MLENativeArchive.__archiveObject(archive).wrapped);
			}
			catch (MLECallError __e)
			{
				throw new SpringMLECallError(__e);
			}
			
			return null;
		}
	},
	
	/**
	 * {@link NativeArchiveShelf#archiveEntry(NativeArchiveBracket,
	 * String)}.
	 */
	ARCHIVE_ENTRY("archiveEntry:" +
		"(Lcc/squirreljme/jvm/mle/brackets/NativeArchiveBracket;" +
		"Ljava/lang/String;)" +
		"Lcc/squirreljme/jvm/mle/brackets/NativeArchiveEntryBracket;")
	{
		/**
		 * {@inheritDoc}
		 * @since 2024/03/05
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			Object archive = __args[0];
			String name = __thread.asNativeObject(String.class, __args[1]);
			
			try
			{
				NativeArchiveEntryBracket entry =
					NativeArchiveShelf.archiveEntry(
						MLENativeArchive.__archiveObject(archive).wrapped,
						name);
				if (entry == null)
					return SpringNullObject.NULL;
				
				return new NativeArchiveEntryObject(__thread.machine,
					entry);
			}
			catch (MLECallError __e)
			{
				throw new SpringMLECallError(__e);
			}
		}
	},
	
	/** {@link NativeArchiveShelf#archiveOpenZip(byte[], int, int)}. */
	ARCHIVE_OPEN_ZIP("archiveOpenZip:" +
		"([BII)Lcc/squirreljme/jvm/mle/brackets/NativeArchiveBracket;")
	{
		/**
		 * {@inheritDoc}
		 * @since 2024/03/05
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			SpringArrayObjectByte buf = (SpringArrayObjectByte)__args[0];
			int off = (Integer)__args[1];
			int len = (Integer)__args[2];
			
			try
			{
				return new NativeArchiveObject(__thread.machine,
					NativeArchiveShelf.archiveOpenZip(buf.array(),
						off, len));
			}
			catch (MLECallError __e)
			{
				throw new SpringMLECallError(__e);
			}
		}
	},
	
	/**
	 * {@link NativeArchiveShelf#entryIsDirectory(NativeArchiveEntryBracket)}.
	 */
	ENTRY_IS_DIRECTORY("entryIsDirectory:" +
		"(Lcc/squirreljme/jvm/mle/brackets/NativeArchiveEntryBracket;)Z")
	{
		/**
		 * {@inheritDoc}
		 * @since 2024/03/05
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			NativeArchiveEntryObject entry =
				MLENativeArchive.__entryObject(__args[0]);
			
			try
			{
				return NativeArchiveShelf.entryIsDirectory(entry.wrapped);
			}
			catch (MLECallError __e)
			{
				throw new SpringMLECallError(__e);
			}
		}
	},
	
	/** {@link NativeArchiveShelf#entryOpen(NativeArchiveEntryBracket)}. */
	ENTRY_OPEN("entryOpen:" +
		"(Lcc/squirreljme/jvm/mle/brackets/NativeArchiveEntryBracket;)" +
		"Ljava/io/InputStream;")
	{
		/**
		 * {@inheritDoc}
		 * @since 2024/03/05
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			NativeArchiveEntryObject entry =
				MLENativeArchive.__entryObject(__args[0]);
			
			try
			{
				return __thread.proxyInputStream(
					NativeArchiveShelf.entryOpen(entry.wrapped));
			}
			catch (MLECallError|IOException __e)
			{
				throw new SpringMLECallError(__e);
			}
		}
	},
	
	/**
	 * {@link NativeArchiveShelf#entryUncompressedSize(
	 * NativeArchiveEntryBracket)}.
	 */
	ENTRY_UNCOMPRESSED_SIZE("entryUncompressedSize:" +
		"(Lcc/squirreljme/jvm/mle/brackets/NativeArchiveEntryBracket;)J")
	{
		/**
		 * {@inheritDoc}
		 * @since 2024/03/05
		 */
		@Override
		public Object handle(SpringThreadWorker __thread, Object... __args)
		{
			NativeArchiveEntryObject entry =
				MLENativeArchive.__entryObject(__args[0]);
			
			try
			{
				return NativeArchiveShelf.entryUncompressedSize(entry.wrapped);
			}
			catch (MLECallError __e)
			{
				throw new SpringMLECallError(__e);
			}
		}
	},
	
	/** End. */
	;
	
	/** The dispatch key. */
	protected final String key;
	
	/**
	 * Initializes the dispatcher info.
	 *
	 * @param __key The key.
	 * @throws NullPointerException On null arguments.
	 * @since 2024/03/05
	 */
	MLENativeArchive(String __key)
		throws NullPointerException
	{
		if (__key == null)
			throw new NullPointerException("NARG");
		
		this.key = __key;
	}
	
	/**
	 * {@inheritDoc}
	 * @since 2024/03/05
	 */
	@Override
	public String key()
	{
		return this.key;
	}
	
	/**
	 * Checks if this is a {@link NativeArchiveObject}.
	 * 
	 * @param __object The object to check.
	 * @return As one if this is one.
	 * @throws SpringMLECallError If this is not one.
	 * @since 2024/03/05
	 */
	static NativeArchiveObject __archiveObject(Object __object)
		throws SpringMLECallError
	{
		if (!(__object instanceof NativeArchiveObject))
			throw new SpringMLECallError("Not a NativeArchiveObject.");
		
		return (NativeArchiveObject)__object; 
	}
	
	/**
	 * Checks if this is a {@link NativeArchiveEntryObject}.
	 * 
	 * @param __object The object to check.
	 * @return As one if this is one.
	 * @throws SpringMLECallError If this is not one.
	 * @since 2024/03/05
	 */
	static NativeArchiveEntryObject __entryObject(Object __object)
		throws SpringMLECallError
	{
		if (!(__object instanceof NativeArchiveEntryObject))
			throw new SpringMLECallError(
				"Not a NativeArchiveEntryObject.");
		
		return (NativeArchiveEntryObject)__object; 
	}
}
